### ISSUE TITLE
1.8.9 Porting

### ISSUE THREAD

**Issue by lehjr** [MachineMuse/MachineMusePowersuits#681]
1.7.10 has been mostly busy work for awhile now, for mods, for Forge, and for players stuck on the sidelines waiting for updates. Aside from the nightmare of the rendering, there are also several dependencies that may not be available. Maybe it's time to start working on the port and disable the modules that have dependencies that can't be met until they become available. I just hope the performance is better than 1.7.10.

**Comment by eyeonus** [comment]
Just curious here, what exactly is the nightmare with rendering?

**Comment by lehjr** [comment]
There are some long winded explanations around but basically it requires a complete rewrite of all the mod's rendering code. The player really won't notice other than the time it takes to get the mod ported, but the developer will pretty much look like this to get it done:

![programmer](https://cloud.githubusercontent.com/assets/1084367/12143321/99cff3c8-b44d-11e5-94db-f79bbe5cf314.gif)

**Comment by lehjr** [comment]
Checked out how bad Numina will be with 1.8.8, so far got it down to 45 errors.

**Comment by flowtender** [comment]
Cool, let's prioritize that for now then. I'll have to fiddle with the build scripts to get a proper separation between 1.7.10 and 1.8.8 builds.

**Comment by lehjr** [comment]
It's the render code that will be the toughest to deal with. I looked at a short block tutorial for 1.8.8 and almost immediately lost my faith in humanity.
Edit:
http://bedrockminer.jimdo.com/modding-tutorials/basic-modding-1-8/model-files/

**Comment by flowtender** [comment]
Yes, I remember begging Grum not to do this.

**Comment by lehjr** [comment]
Just a thought, but how is the player going to charge the armor in 1.8.x?
Edit, looks like there might be some work on Buildcraft for 1.8.x builds, so there must be some sort of power API. 

On a side note, LOL!!
- What went wrong:
  Execution failed for task ':deobfMcSRG'.
  > Your Access Transformers be broke!

I was expecting that, just not worded quite like that.

**Comment by flowtender** [comment]
I'm sure there is some kind of power API in 1.8+.

Regarding rendering, we will have to circumvent the new rendering system rather than using it directly, since it is limited to cubic shapes - pretty much just like before. Does the updated Forge have working wavefront obj support?

**Comment by Xen1290** [comment]
If i remember correctly, in Forge 1.8.9/1.8.8 there is a new rendering system, Made by @RainWarrior.
For power api, the only that exist is the unofficial RF api ported by Progressive Automation devs.

**Comment by flowtender** [comment]
Might be a good opportunity to go power-free instead. Limit modules by power generation of a generator module, instead of batteries.

**Comment by lehjr** [comment]
Yes, it appear according to one of the issues on Forge's github. As far as power goes, an API is one thing, a way to generate it is another.

**Comment by flowtender** [comment]
Alternatively just keep it as-is and let people charge using MPSA generators.

**Comment by lehjr** [comment]
They are very slow to charge.
Edit: maybe change their charge rate until there is stable way to charge them externally?

**Comment by flowtender** [comment]
ok. We could also finally implement a 'storage' module with a UI that contains several slots which are unlockable by different modules:
- solidfuel generator module unlocks a slot that can have fuel put in it and generates energy according to the energy value of the fuel (slider between efficiency and consumption rate)
- subspace energy gate unlocks a slot that can have ender pearls put in them and consumes them semi-randomly (say, 0.001% chance per tick will give the pearls a half-life of about 1hr)
- backpack module unlocks e.g. a 3x5 inventory in which extra items can be stored
- crafting table module unlocks slots in this UI instead of creating a separate UI

This was a planned feature from the start but I got frustrated with minecraft's UI system (obviously) and never got around to it.

**Comment by lehjr** [comment]
Sounds interesting.  Looks like this is going to be quite an overhaul. By the looks of it, none of the modules with external dependencies will carry over for awhile.

**Comment by lehjr** [comment]
Progress so far, LOL!!! : https://gist.github.com/lehjr/1709dfde96643ee8ff2d

That's Numina and MPS in the same workspace.

**Comment by Xen1290** [comment]
Update directly to 1.8.9 if possible

**Comment by lehjr** [comment]
Actually, I'm currently considering falling back to 1.8 and starting from there. 

1.8.9 isn't really an option at this time. With a new version every day, it's too unstable to target, I'm not even sure about 1.8.8 at the moment since several dependencies including NEI are missing.

**Comment by Xen1290** [comment]
I think you can use JEI, every 1.8.9 mod uses It as far as I know, like blood magic, tech reborn and others. I read that is a nei replacement with an easy-to-use API. Maybe @mezz can clarify.

**Comment by mezz** [comment]
1.8.9 should be the current target for mods doing updates. By the time you finish, 1.8 will be completely dead, especially now that Thaumcraft has updated to 1.8.9.
Mods with JEI support in 1.8.9 include Chisels & Bits, Tinker's, Blood Magic, and ExU.
[You can get started with JEI here.](https://github.com/mezz/JustEnoughItems#developing-addons)

**Comment by lehjr** [comment]
Well, at least the MCP mappings for 1.8.9 haven't changed for a few days. Wait, "This set of MCP mappings was designed for MC 1.8.8. Use at your own peril." Um...

**Comment by flowtender** [comment]
Yes, a few other sources have confirmed that JEI+1.8.9 is the standard to shoot for if we can get this done in any reasonable amount of time. Though there is still much uncertainty about power systems.

**Comment by lehjr** [comment]
I don't think it's impossible. I wish my skill level and experience were a little more fitting for the task, but I'll learn as I go, as I always do. My biggest setback is a lack of working examples and documentation, making it hard to figure out how to go from "what was" to "what is". Some of it is cleaner with the use of pos instead of x, y, and z, but some things are quite intimidating. And ffs, it's like someone is getting a commission for every json file needed.

Right now, the only working example I have to work from is this:
https://github.com/MightyPirates/TIS-3D

It does have a couple of things though, like OBJ rendering. 

Edit: for power I'm using the RF API copied from buildcraft's experimental builds.

**Comment by lehjr** [comment]
This is a great read:
https://docs.google.com/document/d/1c8uipS3BOe0BQKH6-kVFCQh8BaXOdRn8SFaAa7HmgWU

**Comment by lehjr** [comment]
So it looks like Buildcraft has an Alpha build for 1.8.9. That should be a good start for getting power. 

For 1.8.9 I think it would be best to add "optional recipes" so that the default recipes are overwritten on update. That will address the insanity of having to constantly address recipe issues have already been fixed.

**Comment by lehjr** [comment]
Sort of have something rendering, too bad it's stuck to the player:
![2016-01-14_12 29 03](https://cloud.githubusercontent.com/assets/1084367/12332102/a8a90204-baba-11e5-8169-ed70b4d9fd57.png)

**Comment by flowtender** [comment]
ahahaha adorable :>

**Comment by lehjr** [comment]
Sadly, I'm still looking at what to use for rendering the TinkerTable. Since IBSR's were removed in 1.8 I used a TESR just to see it working in the current workspace. I'm not entirely sure if I can avoid using one, but I'm looking at alternatives.

**Comment by lehjr** [comment]
The armor models will be the next challenge. I'm reasonably confident that will be easier. On a side note, the 1.8.9 mods seed to have more than doubled in the past week.

**Comment by lehjr** [comment]
Well, that took longer than it should have, but I did gain some knowledge from it. 
![2016-01-15_15 33 07](https://cloud.githubusercontent.com/assets/1084367/12364658/eef99adc-bb9d-11e5-8a38-63d527b48c09.png)

**Comment by eyeonus** [comment]
I was going to say it looks a little blocky, but then I thought you might not get the joke.

**Comment by lehjr** [comment]
LOL

**Comment by lehjr** [comment]
@MachineMuse which side if the TinkerTable is the front or does it really matter?

**Comment by flowtender** [comment]
It doesn't matter. In fact in 1.7 and earlier it could not be rotated at all because it was the very first block I made and it was symmetrical on 4 axes before I replaced it with SebK's model. I never went back and made it properly rotation-sensitive.

**Comment by lehjr** [comment]
Making some progress, but still have issues to work around. There are lots of changes, some small, some large. Some stuff I have no idea what to do with, like this:
https://github.com/MachineMuse/Numina/blob/experimental/src/main/scala/net/machinemuse/numina/render/MuseIconUtils.scala#L62-L65
because `addVertexWithUV` no longer exists I've yet to find a replacement.
There are other fun things as well, like OBJ rendering (currently I'm using the old OBJ setup that's in Thaumcraft 5). I have about 50 or so errors to contend with just to get things to where they will build. Running might be another story :P

**Comment by RainWarrior** [comment]
Please don't do what TC5 did with OBJs. There's a loader in forge now, use it.

**Comment by lehjr** [comment]
No worries. I know the old rendering was removed for a reason and I'm only using the old setup just to clean up everything else to where it will build. After that I'll dig into the new setup and do things right.

**Comment by AlexIIL** [comment]
Instead of using the tessellator to add vertexes, you use

``` java
WorldRenderer wr = tess.getWorldRenderer();
// There are several to choose from, look at DefaultVertexFormats for more info
wr.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX);
// Imaginary class Point that contains the position and texture coordinates
for (Point p : points) {
  wr.pos(p.x, p.y, p.z);
  wr.tex(p.u, p.v);
  // This is very important
  wr.endVertex();
}
tess.draw();
```

One thing you need to make sure is that you call pos(), tex(), normal() and color() in the same order as they are defined in the vertex format you are using.

**Comment by lehjr** [comment]
Thanks!!!

**Comment by flowtender** [comment]
by the way, a TESR is fine. sorry, should have mentioned that earlier.it's not a block that will be in worldgen, and most people aren't making buildings out of them, so it's OK to have them as tile entities with no stored information (or they can be tile entities with information stored about orientation, nbd either way)

**Comment by lehjr** [comment]
about 40 errors to go for a successful build. Then I need to figure out the new rendering for OBJ.

**Comment by lehjr** [comment]
Took the day off yesterday to build a new Linux kernel (yes people still do that) and update my video drivers, so not much progress atm. However, I am quite certain I am going to reach a point where I will need help because some of this is quite hard to follow, some of it I just have no idea what to do with it. I'll be uploading what I have so far later today, but keep in mind that is an absolute mess.

**Comment by flowtender** [comment]
okay :> that's super cool anyway. No pressure.

**Comment by lehjr** [comment]
I still have a lot more to upload, but I'm holding off on that while I experiment a bit with the armor rendering in a lighter/less complex setup since the rendering itself will be a bit for me to figure out.

**Comment by lehjr** [comment]
I guess this could be a fallback option until I figure out the rendering:
![2016-01-28_15 23 02](https://cloud.githubusercontent.com/assets/1084367/12657785/3c9e4f50-c5d3-11e5-8c87-ebac2f5121b9.png)

**Comment by flowtender** [comment]
I'm sure a vocal minority of users will be very happy about that, lehjr :)

**Comment by flowtender** [comment]
We don't have to update the custom rendering if it's too much effort. It was entirely a cosmetic feature, deliberately divorced from the functionality of the mod.

**Comment by lehjr** [comment]
It's too early for me to say for sure if it's too much effort to get working. The only big issue I have is lack of examples, but I'm going to give it a week and see if I can figure it out. Everything that isn't model or texture related looks easy.

**Comment by Razorskills** [comment]
Sometimes I wish there was a config to switch it to this armor instead of the new one, just for nostalgia purposes.

**Comment by flowtender** [comment]
Yes, it has been requested in the past. I think I closed it once during a period of grumpiness, but it's a reasonable request.

**Comment by lehjr** [comment]
Right now I can't tell how feasible it is to implement. Lots of things have changed, like json files to get working icons, or the loss of direct access block textures.
For instance `Blocks.water.getIcon(0, 0);` now would be something like:
`Minecraft.getMinecraft().getRenderItem().getItemModelMesher().getItemModel(new ItemStack(Blocks.water)).getParticleTexture()` followed by a few mixed drinks.

**Comment by lehjr** [comment]
Progress is slow, but steady, kind of like a herd of turtles. It's a deceptively complex mod. I'm finally at a point where I can run the mod from the real source rather than just a basic tutorial version, but the only thing rendering at this point is the Tinker Table, and only when placed, not as an inventory item. 

Edit: Tinker table GUI partially works again.

**Comment by flowtender** [comment]
^^' Smoke and mirrors

**Comment by lehjr** [comment]
Meters are now working. More of the Tinker table GUI is working except module icons are not yet working and swirly circles are partially working. Not as much progress as I would have liked, but at least it's something.

**Comment by flowtender** [comment]
All good :) at least it compiles! Ship it!

**Comment by lehjr** [comment]
Sliders now work in the Tinker table GUI. Power seems to be working. Armor models haven't been touched yet. The power fist is still untouched. I have no idea what modules are working yet. Anything released at this point would have to be labeled as an early alpha.

**Comment by flowtender** [comment]
Sure! Do you want to push it to a different branch and I'll label it as such?

**Comment by Xen1290** [comment]
I suggest to you a 1.8.9 branch in this repo, and maybe build and release automatically as alpha in curseforge when you release or tag a version in github, like applied energistics does.
I hope I explained well.
P.S. The official RF 1.8.9 api is released. Take a look here: https://github.com/CoFH/RedstoneFlux-API/tree/1.8

**Comment by flowtender** [comment]
Yeah, that's what I'm planning to do.

**Comment by lehjr** [comment]
Yeah, I would do a 1.8.9-alpha branch and rename it later when the the mod is in better shape. Just keep in mind that Numina source is currently mixed in with MPS.

**Comment by flowtender** [comment]
ok.

**Comment by lehjr** [comment]
Once I get the rendering figured out I'll split the source back up

**Comment by lehjr** [comment]
@Xen1290 Thanks for pointing out that RF api release. I was hoping that would turn up.

**Comment by Xen1290** [comment]
👍🏻

**Comment by lehjr** [comment]
Started working on getting module icons to load... good times. Good times. I should be done with that much tonight if all goes well.

**Comment by lehjr** [comment]
Or not... wow, so many modules, so many icons. All but the tool module icons done now. 

Edit:
 All but 2 working now.

**Comment by lehjr** [comment]
Tinker Table with different model systems. The top half of the screen is using  iChun's Tabula, the bottom is using MrCrayfish's Model Creator. I also did one with Techne

![screenshot from 2016-02-20 13 53 49](https://cloud.githubusercontent.com/assets/1084367/13198539/8d7ce978-d7d9-11e5-97a7-b2b1535ee8d1.png)

**Comment by Xen1290** [comment]
If possibile, don't use techne.

**Comment by lehjr** [comment]
The problem with all of it is item rendering which is why I wanted to hold off on 1.8.9 in the first place and go with 1.8, which in hindsight, I would have been finished with by now. It seems that item rendering for anything that doesn't just use an icon instead of a model is a pain in the ass.

Edit: I think I'm going to try to get the json model right and use that. It will take more time than I'd like but it should be future-proof for a few MC versions.

**Comment by lehjr** [comment]
There, we now have a Tinker Table json model for rendering it as an item. It's not quite good enough for use as a block model because lining up the textures was a pain. Even with precision like this:
`"east":  {"uv": [13.7142857142857, 3.42857142857143, 14.2857142857143, 4], "texture": "#0"}` 
it's still not quite right but still good enough for an item model. 

Anyway, now for the Power Fist. Also, I need some kind of idea for version numbers for 1.8.9 before I upload anything.

**Comment by Xen1290** [comment]
Alpha: 0.0.x.jenkinsprogressivenumber
Beta: 0.x.y.jenkinsprogressivenumber
Release: x.y.z.jenkinsprogressivenumber

**Comment by Xen1290** [comment]
So, will be ModularPowersuits-1.8.9-0.0.1.1.jar for the first jenkins and curse build

**Comment by lehjr** [comment]
@Xen1290 Great thanks! The first couple of builds may not be jenkins builds but as long as I have the versions numbers set up correctly it should be OK later on. Actually, I have a build ready for upload but I wanted to make sure the version numbers were set up correctly. I still have to break Numina back out into its own source set, but it shouldn't be too difficult using Meld and comparing with a 1.7.10 source set.

**Comment by Xen1290** [comment]
I think is the best version number set. Because now we don't have the minecraft version and start from 0.11 is strange.
So, new version, new number version.

**Comment by flowtender** [comment]
I was going with just incrementing the version number regardless of new Minecraft versions based on the size of changes (revision = fixes, minor = new features, major = someday a release will come~)

The major version number was kept at 0 so fewer people expect a hobby project to act like a fully released and functioning product.

**Comment by lehjr** [comment]
Oops. Well, the first release is up and I reset the counter. Still working on the mod as much as I can. After I get everything more or less working I have to go back and fix the power fist model. The current model was to avoid having to do something like using a Minecraft stick and renaming the device to "power stick" :P

Also the scaling for OBJ stuff is a pain. Lux capacitor model is insanely huge atm.

**Comment by flowtender** [comment]
It should be a factor of 16 I think. I kind of worked around the weird scaling by having the render subsystem scale the models individually. I guess that's not an option with Minecraft's new VERY MUCH BETTER AND NOT AT ALL HEADACHE INDUCING format.

**Comment by lehjr** [comment]
No, I mean setting the scale from the json file that loads it doesn't seem to do anything.

**Comment by flowtender** [comment]
...hahahaha :(

**Comment by lehjr** [comment]
I'll figure it it out, I think. The blade launcher works again, the plasma cannon is having issues with the bolt size. There might be an issue between the keyboard and the chair. Not sure.

**Comment by TextureTurtle** [comment]
i have made two different possible armor skins
![unnamed 1](https://cloud.githubusercontent.com/assets/16947391/13272024/14f58588-da67-11e5-96dc-32089606d02e.png) and 
![unnamed](https://cloud.githubusercontent.com/assets/16947391/13272034/24d20db4-da67-11e5-867a-d0159e1f6859.png)

**Comment by lehjr** [comment]
With any luck, players will be able to use custom skins by setting the mod config settings to vanilla models and overriding the default skin with their own with the use of resource packs. I don't know if there is a way for players to be able to see each other's custom skin yet though.

**Comment by flowtender** [comment]
There is a framework there in the config uploader/downloader. I never got around to fully implementing it since I was focused on numina/anima stuff by then.

**Comment by lehjr** [comment]
There, got the Lux Capacitor throwable entity to launch with the obj model. Just need to get the TESR part working now. After that, I'll tackle the armor rendering. Not sure when, but later on I'd like to see what I can do with getting the armor to toggle between the vanilla model and the OBJ model so users can use their own skins.

**Comment by lehjr** [comment]
ran into this today: https://github.com/MinecraftForge/ForgeGradle/issues/344
good times :S

**Comment by lehjr** [comment]
I feel like I'm reinventing the wheel. 

![screenshot from 2016-03-01 11 26 15](https://cloud.githubusercontent.com/assets/1084367/13439533/0a65cc68-dfbc-11e5-8f2e-1947114d9aca.png)

**Comment by flowtender** [comment]
Y-you don't have to do that D:

**Comment by lehjr** [comment]
Just looking for a way to render an item model. IItemRenderer is gone and there is no replacement.

**Comment by flowtender** [comment]
:<

**Comment by lehjr** [comment]
Hoping to have the next alpha ready by the end of next week. Things scheduled in the next build are:
1. completely working blade launcher (done - texture was missing in previous build)
2. completely working lux capacitor ( 50% done. Recreated the model in json format after headaches with the OBJ loader. Currently working on rotations and coloring without using TESR.)
3. fix pickaxe module (not started - there is a major bug/porting failure in handleHarvestCheck. I just need to edit it and figure out what information I really need to get out of it to pass to the modules)
4. switch to vanilla armor model and disable customization screen for armor (started. Switch to an old armor skin for now. still need to disable the customization screen. getArmorModel is not getting the correct armor slot for armor pieces, might be a forge bug)

The OBJ loader seems pretty fragile and I had issues getting it to load the models.

**Comment by lehjr** [comment]
Fixed packet bug that crept in during porting.

**Comment by lehjr** [comment]
Looks like I'll be pushing the next alpha out later today. The shovel module is still a bit buggy because the harvest check event doesn't have any easy way to get the coordinates that you need to get the blockstate which is needed for the harvest check, specifically for getHarvestTool(IBlockState state) which is used to check the module to see if it is emulating the tool needed to harvest the block. 

Anyway. I will be disabling the Lux Capacitor for now until I can get it to work like it did in 1.7.10. Most of the modules are working now. Mostly what is left is rendering stuff which I will start working on this week.

**Comment by Xen1290** [comment]
Great work @lehjr ! Keep it up!

**Comment by lehjr** [comment]
Looks like it might be time to target 1.9, or at least try

**Comment by Xen1290** [comment]
Yeah, every mod will update to 1.9.

**Comment by flowtender** [comment]
If you need/want to change some functionality in order to fit with new minecraft versions, go right ahead. It doesn't have to be a perfect clone. :P

**Comment by lehjr** [comment]
The biggest issue I have is the rendering. I really need to set some "quiet time" aside and just figure out how the rendering works and see what I need to do to get things working. The models will need to be cleaned up a bit to get rid of warnings about unsupported keys or w/e. I imagine Blender will work, and I do have it, but the last time I did anything in 3D was with DesignCAD 3D, back when PC's had "Turbo" buttons.

**Comment by eyeonus** [comment]
If you can link me the programs/files I need for doing modeling, I can take
a crack at it. I'm pretty familiar with Blender.

Just don't ask me to do any coding.

**Comment by lehjr** [comment]
@aionys Basically, it's all of the Wavefront models here: https://github.com/MachineMuse/MachineMusePowersuits/tree/experimental/src/main/resources/assets/powersuits/models

The other issue is the texture files now have to be a multiple of 16 and have to be square. I also have a wavefront model for the power fist, somewhere.

**Comment by eyeonus** [comment]
Okay.... What exactly do you need done? They load perfectly fine in
Blender.....

On Tue, Mar 22, 2016 at 6:53 AM, lehjr notifications@github.com wrote:

> @aionys https://github.com/aionys Basically, it's all of the Wavefront
> models here:
> https://github.com/MachineMuse/MachineMusePowersuits/tree/experimental/src/main/resources/assets/powersuits/models
> 
> The other issue is the texture files now have to be a multiple of 16 and
> have to be square. I also have a wavefront model for the power fist,
> somewhere.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/MachineMuse/MachineMusePowersuits/issues/681#issuecomment-199797737

## 

mu'o mi'e .aionys.

.i.e'ucai ko cmima lo pilno be denpa bu .i doi.luk. mi patfu do zo'o
(Come to the Dot Side! Luke, I am your father. :D )

**Comment by lehjr** [comment]
@aionys yeah. They load fine in Blender, but in Minecraft, the amount of log spam generated just isn't going to work. There's no way to silence warnings about invalid keys. It might be something as simple as importing, and exporting with certain settings, I'm not sure. The other issue is as I said. The textures now have to be square and multiples of 16, which throws off the mappings. ATM, I don't have logs for that and I'm currently tracking down another issue, so it may be a few days before I can get the models to load to where I can even get the logs for that

**Comment by lehjr** [comment]
There, finally fixed the plasma bolt. Now to focus on armor rendering.

**Comment by lehjr** [comment]
This is what I was talking about with the wall of warnings: https://gist.github.com/lehjr/d1acec7f03bd1d3e62f6

**Comment by lehjr** [comment]
decided to hold back on 1.9 at least until I can get 1.8.9 to a beta state. I still need to write a JEI recipe handler (in progress), fix the Lux Capacitor, and fix the power fist rendering (maybe use item entity?) first. I'm not entirely convinced the current OBJ model rendering will work for the armor since groups seem to be tied to blockstates. The JEI recipe handler is the priority at the moment, since it seems players want to be able to craft items.

**Comment by Xen1290** [comment]
@lehjr If you need help with JEI, I think that @mezz is the right guy.

**Comment by lehjr** [comment]
finally got enough quiet time to get the recipe handler working. :D

**Comment by lehjr** [comment]
I'm gonna push out the update for Numina with the recipe handler this weekend. In the mean time I want to see what I can do with the model rendering. The big thing is to resize the textures without hosing the UV mapping.

Edit: looks like as long as the textures are re sized by stretching, the mapping should stay the same for the OBJ models.

**Comment by lehjr** [comment]
I'll be working on a 1.9.4 port in the next couple weeks. From a gameplay perspective, the biggest issue of the moment is the lack of high end power generators, making charging anything but the most basic power level nearly impossible. For 1.8.9, I'm using Advanced Generators and Pressure pipes. 

From a development standpoint, I'm going to need to fix a couple of things first, like the recipe handler which is missing the ability to place the items in the crafting interface as well as fixing the handling of recipes. There's also a weird behavior where if the first line of the recipe doesn't have all 3 placeholders used, the recipe displayed is rotated and invalid. 

The model issues will likely be addressed in 1.9.4. The power fist model needs to be fixed. I don't know if we'll ever get animated item models back again though. The other thing I'm not sure about is the customization for the armor models like we had in 1.7.10. I'm just not even sure how to implement it. 

Also, on the Lux Capacitor. I'm thinking about getting rid of the custom color for it and maybe just having a couple preset colors. In doing so, we can completely eliminate using a TE and a TESR, allowing these to be useful for base and tunnel lighting without really effecting performance. I'm also considering changing the recipe setup a bit so that there are both builtin/default recipes that will get overwritten on loading and optional additional custom recipes to help eliminate the bug reports for recipes that have already been fixed but don't get overwritten.

**Comment by Lordmau5** [comment]
> Also, on the Lux Capacitor. I'm thinking about getting rid of the custom color for it and maybe just having a couple preset colors. In doing so, we can completely eliminate using a TE and a TESR, allowing these to be useful for base and tunnel lighting without really effecting performance.

Or you can have a config option to allow both variants.
The FastTESR should not make a huge impact on those, as far as I know (well, it strongly depends on how many lights you place down in the end)

Technically I'm rendering (5x5x5 - 3x3x3) FastTESR models on one tank in my mod as well and it doesn't cause any visible lag.

Still, I'd like to page @RainWarrior on this, since he's a bit more experienced with rendering and that stuff, iirc :P

**Comment by lehjr** [comment]
I did an experiment with the current setup and it took about 50 or so before I started seeing lag on my PC. A higher end video card would take more, but the point is better optimization. Currently the only thing the TESR is being used for is the colored lens, the model itself is a JSON model, but it doesn't always work. And quite frankly, when it comes to the rendering code, I'm not just horrible at it, I'm short-busing it all the way.

**Comment by Lordmau5** [comment]
Is the newest code on your Github as we speak?
I wouldn't mind taking a crack at looking over it (and perhaps even going as far as hitting it up in a 1.9.4 repo)

**Comment by eyeonus** [comment]
*affecting performance

As far as limited colour options are concerned, if you decide to do that, I'd suggest going with the standard 16, possibly excluding black.

**Comment by lehjr** [comment]
Be warned, this code is such a mess that it may cause paralysis, temporary blindness, and insanity (no really, I'm not exaggerating): 

MPS code here: https://github.com/lehjr/MachineMusePowersuits/tree/1.8.9
Numina code here: https://github.com/lehjr/Numina

**Comment by lirannl** [comment]
If an issue of yours with 1.9.4 is power, there's EnderIO. I absolutely LOVE that mod, it has vibrant capacitors, and vibrant energy conduits transferring up to 20k RF/t. High end enough?

By the way, EnderIO did some fancy stuff with rendering. If that helps in any way.

**Comment by lehjr** [comment]
I love EnderIO, but a fully charged power suit with maxed out power storage can hold millions of RF per piece and it would take several top end generators and days of charging. My last setup in 1.7.10 was using a Mekanism fusion reactor. My current setup in 1.8.9 is using Advanced Generators Steam Generator with bronze turbines and 2 heat exchangers with lava pumping from the nether and an EnderIO wireless charger.

**Comment by Xen1290** [comment]
@lehjr maybe you can use the creative capacitor bank from EnderIO or debug item+MFSU from IC2 to charge items.

**Comment by lehjr** [comment]
@Xen1290 for power I was referring to actual gameplay as a regular player would play. Without the ability to charge the armor, it becomes useless rather quickly and likely would be unused, and not generating the feedback I need to find and fix issues. I just took a couple weeks off from working on the mod, to work with the mod, playing with it as an actual player to find and fix issues. The recipe handler was horrible and there were broken plating recipes that I didn't know about. I also found the tooltips for the power are missing. I fixed the recipes and the handler, but the broken tooltips are still on my todo list. The hoe module also seems a bit off (it's supposed to be able to do more than a 1x1 square). I'll probably fix those last 2 issues before moving to 1.9.4.

**Comment by Razorskills** [comment]
If you you are thinking of making your own armor charger, maybe an armor stand that you could right click and have it transfer armor and power tool onto it, maybe burn coal, or hook up your own generator to it. But also allow rf pipes to hook up from an external mod.

**Comment by lehjr** [comment]
No, I'm merely pointing out that the top tier power modules require more than simple generators and that without those, the mod receives less usage than needed in order to get the bugs worked out. There are MPS modules for charging, but they are quite underwhelming. If I were to add any further methods of charging, it would likely be through modules as adding external devices doesn't fit well into the theme of the mod. An armor stand that does charging would be better as a stand alone mod as there are other armors out there that use power. 

The problem with implementing high tier charging in MPS is balance and mod interaction. If you add an internal method of charging at the highest level possible, then those high end power mods really don't have as much draw to them. Not only that, but the armor itself could end up being a source of infinite power. It would really nice to implement the EV Capacitor, which would become the highest tier power system, but again balance and power issues.

Charging is much more problematic than you would imagine. In fact, there is a partially implemented coal generator in the code, but in order to make it balanced, it needs to put out the same amount of charge as a stand-alone generator, which isn't much. Keep in mind though that you can also use charged devices as power sources by carrying them in your inventory, such as EnderIO capacitor banks and various devices from IC2.

**Comment by Xen1290** [comment]
Oh sorry, now I understand what you mean

**Comment by lirannl** [comment]
I still don't get one thing - is enderIO's Vibrant tier too low for you?

**Comment by lehjr** [comment]
@lirannl there's no high tier generators to go with them. Right now, in my actual gaming setup, the Advanced Generator setup is producing over 10k RF/tick. EnderIO has no such power generation. Yes, you can use capacitor banks to store power, but you still need a way to charge them in a reasonable amount of time.

**Comment by lehjr** [comment]
Top tier MPS armor piece stores 50 million RF for each piece. So the question becomes how do you charge 250 Million RF in a reasonable amount of time.

Edit: I mean without using top tier power generation mods.

**Comment by Lordmau5** [comment]
I think when you have _"Top tier MPS armor"_, you also have _"top tier power generation"_ O_o

//EDIT: Also, one Endergenic Generator setup from RFTools can yield to ~25k RF/t or even more.
Besides taking in Ender Pearls it's not that expensive (heard this from my friend btw)

//EDIT 2: Taking 250m RF into account for the full armor, + 25k RF/t from the generator:
25k RF/t \* 20 => 500k RF/s
250 000 000 RF / 500 000 RF/s => 500s

So around 8 minutes.

**Comment by lehjr** [comment]
Not without relying on mods that may not be out. With "Top tier MPS armor" you have top tier power consumption and storage capability, but very **low tier power generation** capabilities in the mod itself.

**Comment by lehjr** [comment]
@Lordmau5 right, but again, that would be considered top tier generation rather than something like EnderIO's Stirling Generator at around 20-60 RF/t where it would not be considered a reasonable about of time.

**Comment by lirannl** [comment]
@lehjr
I could just go overkill with the sitrling generators. I could do crazy stuff with mana and convert it to RF. Leave filling up a vibrant capacitor bank to the player. The point is, a full vibrant capacitor (or a few, or a wireless charger hooked up to a power grid) can definitely charge a power suit in a timely manner. As I said, the player will have mods to generate RF, and could always go nuts and build a LOT of generators or high tier ones (what if extra utilities 2 reimplements the extra utilities generators?). Stop worrying about that, the vast majority of players use mod packs, which will likely contain enderIO and Rftools, or some other way of generating a lot of rf.

**Comment by lehjr** [comment]
> I could just go overkill with the sitrling generators

Sure, at a max of 60 RF/tick you'll need at least 250 of them to make it worth while.

> what if extra utilities 2 reimplements the extra utilities generators

The power output wasn't enormous. Fuel consumption scaled with the output and at times became impractical. Even if they come back, it's not likely they'll change the way you charge the armor. 

> ... which will likely contain enderIO and Rftools, or some other way of generating a lot of rf..

Such as? EnderIO has no way of generating a lot of RF since it doesn't need it. I'm still not seeing all that many other options yet. Maybe Mekanism if/when it's done for 1.9.4. 

> I could do crazy stuff with mana and convert it to RF

or maybe I do some crazy stuff and charge the armor directly with mana :P (probably not, but it might be an option)

>  Stop worrying about that...

I actually do HAVE to worry about such things as being able to charge the armor. This isn't really a situation where "if you build it, they will come". Things have to be done in a practical manner or there really isn't much point. 

Anyway, it's possible that I could implement the EV capacitor which  would become the new top tier power storage and then re-scale the lower tiers while not really raising the ceiling much higher. I don't know yet. It may or may not really be worth while to implement.
